### PR TITLE
IGNITE-2779 BinaryMarshaller caches must be cleaned during client reconnect.

### PR DESCRIPTION
**Pull request by NSAmelchev**

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/internal/IgniteClientReconnectBinaryContexTest.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        });
+
+        ignites.iterator().next().createCache(personCache);
+
+        client.cache(DEFAULT_CACHE_NAME).put(new TestClass1("2"), new TestClass1("2"));
+        client.cache(DEFAULT_CACHE_NAME).put(new TestClass1("3"), new TestClass1("3"));
+
+        int size = 0;
+
+        for (Cache.Entry<TestClass1, TestClass1> entry : client.<TestClass1, TestClass1>cache(DEFAULT_CACHE_NAME))
```
Comment (daradurvs) [inline modules/core/src/test/java/org/apache/ignite/internal/IgniteClientReconnectBinaryContexTest.java]: assertEquals(2, client.cache(DEFAULT_CACHE_NAME).size() or localSize() it depends on what you need.

File: modules/core/src/test/java/org/apache/ignite/internal/IgniteClientReconnectBinaryContexTest.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+        Ignite srv = grid(0);
+
+        CacheConfiguration<TestClass1, TestClass1> personCache = new CacheConfiguration<TestClass1, TestClass1>()
+            .setName(DEFAULT_CACHE_NAME)
+            .setCacheMode(CacheMode.REPLICATED);
+
+        srv.createCache(personCache);
+
+        client.cache(DEFAULT_CACHE_NAME).put(new TestClass1("1"), new TestClass1("1"));
```
Comment (daradurvs) [inline modules/core/src/test/java/org/apache/ignite/internal/IgniteClientReconnectBinaryContexTest.java]: Let's call it once and use it for the next steps:
IgniteCache cache = client.cache(DEFAULT_CACHE_NAME);

File: modules/core/src/test/java/org/apache/ignite/internal/IgniteClientReconnectBinaryContexTest.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+     * @throws Exception If failed.
+     */
+    public void testBinaryContexReconnectClusterRestart() throws Exception {
+        Ignite client = grid(serverCount());
+
+        assertTrue(client.cluster().localNode().isClient());
+
+        Ignite srv = grid(0);
+
+        CacheConfiguration<TestClass1, TestClass1> personCache = new CacheConfiguration<TestClass1, TestClass1>()
```
[the remark this answers is not in the document]
Comment (NSAmelchev) [reply modules/core/src/test/java/org/apache/ignite/internal/IgniteClientReconnectBinaryContexTest.java]: We have to use user types.

File: modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryContext.java [hunk 4]
```diff
@@ -1339,6 +1339,15 @@ public void unregisterBinarySchemas() {
     }
 
     /**
+     * Unregister user type descriptors.
+     **/
+    public void unregisterUserTypeDescriptors() {
```
[the remark this answers is not in the document]
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryContext.java]: Yes, I missed the iterator

File: modules/core/src/test/java/org/apache/ignite/internal/IgniteClientReconnectBinaryContexTest.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        cfg.setPeerClassLoadingEnabled(false);
+
+        return cfg;
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testBinaryContexReconnectClusterRestart() throws Exception {
+        Ignite client = grid(serverCount());
```
[the remark this answers is not in the document]
Comment (NSAmelchev) [reply modules/core/src/test/java/org/apache/ignite/internal/IgniteClientReconnectBinaryContexTest.java]: I did similar with IgniteClientReconnect...tests

File: modules/core/src/test/java/org/apache/ignite/internal/IgniteClientReconnectBinaryContexTest.java (hunk 1 above)
Comment (NSAmelchev) [reply modules/core/src/test/java/org/apache/ignite/internal/IgniteClientReconnectBinaryContexTest.java]: I need to read each entry to reproduce a bug.

File: modules/core/src/test/java/org/apache/ignite/internal/IgniteClientReconnectBinaryContexTest.java (hunk 2 above)
Comment (NSAmelchev) [reply modules/core/src/test/java/org/apache/ignite/internal/IgniteClientReconnectBinaryContexTest.java]: I need to get cache this way to reproduce a bug. You suggest a way that throws an exception like IllegalStateException.

**Pull request closed without merging**